File: tests/07_phases_ompss_2.dg/fortran/success_task_outline_reduction_01.f90
! <testinfo>
! test_generator=config/mercurium-ompss-2
! </testinfo>

! Reductions arrays outline 1

!$OSS TASK REDUCTION(+: P)
SUBROUTINE S(P, I)
    IMPLICIT NONE
    INTEGER :: P(10)
    INTEGER :: I

    P(I) = P(I) + 1

END SUBROUTINE S

PROGRAM MAIN
    IMPLICIT NONE

    INTEGER :: P(10)
    P = 0
    CALL S(P, 1)
    CALL S(P, 2)
    CALL S(P, 3)
    CALL S(P, 4)
    CALL S(P, 5)
    CALL S(P, 6)
    CALL S(P, 7)
    CALL S(P, 8)
    CALL S(P, 9)
    CALL S(P, 10)
    !$OSS TASKWAIT

    IF ( ANY(P /= 1) ) STOP 1

END PROGRAM MAIN

File: tests/07_phases_ompss_2.dg/c/success_task_outline_reduction_01.c
/*
<testinfo>
test_generator="config/mercurium-nanos6-regions"
</testinfo>
*/

#include <assert.h>

#pragma oss task reduction(+: [1]p)
void bar(int *p) {
    (*p)++;
}

#pragma oss task weakreduction(+: [10]array)
void foo(int *array) {
    for (int i = 0; i < 10; ++i) {
        bar(&array[i]);
    }
}

int main() {
    int array[10];
    for (int i = 0; i < 10; ++i) {
        array[i] = 0;
    }

    foo(array);

    #pragma oss taskwait

    for (int i = 0; i < 10; ++i) {
        assert(array[i] == 1);
    }
}
